src/match-path-async.ts

```diff
@@ -125,48 +167,45 @@ function findFirstExistingPath(
       if (err) {
         return doneCallback(err);
       }
-      const mainFieldName = getPrioritizedMainFieldName(
-        packageJson,
-        mainFields
-      );
-      if (mainFieldName) {
-        const file = path.join(
-          path.dirname(tryPath.path),
-          packageJson[mainFieldName]
-        );
-        fileExists(file, (err2, exists) => {
-          if (err2) {
-            return doneCallback(err2);
-          }
-          if (exists) {
-            // Not sure why we don't just return the full path? Why strip it?
-            return doneCallback(undefined, Filesystem.removeExtension(file));
-          }
-          // Continue with the next path
-          return findFirstExistingPath(
-            tryPaths,
-            readJson,
-            fileExists,
-            doneCallback,
-            index + 1,
-            mainFields
-          );
-        });
-      } else {
-        // This is async code, we need to return in an else-branch otherwise the code still falls through and keeps recursing.
-        // While this might work in general, libraries that use neo-async like Webpack will actually not allow you to call the same callback twice.
-        // An example of where this caused issues: https://github.com/dividab/tsconfig-paths-webpack-plugin/issues/11
-
-        // Continue with the next path
-        return findFirstExistingPath(
-          tryPaths,
-          readJson,
+      if (packageJson) {
+        return findFirstExistingMainFieldMappedFile(
+          packageJson,
+          mainFields,
+          tryPath.path,
           fileExists,
-          doneCallback,
-          index + 1,
-          mainFields
+          (mainFieldErr?: Error, mainFieldMappedFile?: string) => {
+            if (mainFieldErr) {
+              return doneCallback(mainFieldErr);
+            }
+            if (mainFieldMappedFile) {
+              // Not sure why we don't just return the full path? Why strip it?
+              return doneCallback(
+                undefined,
+                Filesystem.removeExtension(mainFieldMappedFile)
```

Comment:
As the comment above asks, why are we actually removing the extension?

Reply:
IIRC I added this comment some time ago when doing the initial work on the async version. @Jontem do you remember why the stripping is done? Or do you also think we can just return the raw filename?

Reply:
The code has been rewritten many times since i looked at it and i don't recall the reason.

Reply:
@christoffer I think you can go ahead and remove the stripping of the extension.

Reply:
@jonaskello Hi Jonas. Sorry for the delay, PTOs and what-not came in the way.

Would you be OK if we remove the extension in a separate diff? I feel that it's unrelated to the change I'm introducing here. And since it *might* have side effects, it would be nice to keep that change separate.

Reply:
My original comment was probably unclear. I was only asking out of curiosity, not as a way to ask if I should remove it now :)

Reply:
@christoffer Sure, doing it in a separate diff makes sense.